Unit-test try_evaluate_pointer_comparison and bugfix

romainbrenguier opened the pull request:
This adds some unit test for `try_evaluate_pointer_comparison`, then fix a small bug, and add an unit-test for the case that was buggy, then add some checks for an implicit invariant that I noticed while unit testing, and make value_sett::output more usable in particular for debugging unit tests. 

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Replace use of get_l1_object by remove_level_2

get_l1_object suggests we are comparing the address of the object, but
the actually should interpret the symbol itself as a pointer.
A unit test for the case of the issue that is fixed will be added in a
following commit.

romainbrenguier committed:
Declare try_evaluate_pointer_comparisons in header

This is so that we can unit-test this function.

romainbrenguier committed:
Unit test for try_evaluate_pointer_comparisions

romainbrenguier committed:
Check pointer evaluation in the member case

Field sensitivity the value set to keep track of pointers in individal
fields, making it possible to evaluate equalities such as the one in the
example.
This is a case which would not work if try_evaluate_pointer_comparisons
was using get_l1_object_identifier instead of getting the identifier of
the whole l1 expression (not the underlying object).

fixup! Check pointer evaluation in the array case

smowton commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -1220,7 +1221,8 @@ void value_sett::assign(
   output(ns, std::cout);
 #endif
-
+  PRECONDITION(is_l1_renamed(lhs));
```
This is only true when used in symex-- otherwise we expect to operate on plain un-renamed expressions. Therefore these invariants should move to `symex_dereference` or to some wrapper function in `goto_statet`.

romainbrenguier replied:
I defined a wrapper class for `value_sett` (a wrapper function wouldn't ensure we didn't continue to use the unchecked one)

owen-mc-diffblue reviewed (approved):
Great unit tests. Some other edge cases you might want to consider testing: if the value-set contains `ID_unknown`, `ID_invalid` or a failed symbol then `try_evaluate_pointer_comparison` should not be able to determine anything.

romainbrenguier committed:
Add case to try_evaluate_pointer_comparisons test

This test the case where ptr1 can point to value1 or value2 and we
evaluate `ptr1 != &value1`

romainbrenguier committed:
Add case for evaluate pointer comparison with null

This check the case where ptr1 is either &value1 or &value2 and we
compare it to null.

romainbrenguier committed:
Test try_evaluate_pointer_comparisons unknown case

This adds a unit test for the case where the value set contains
unknown. In that case no comparison can be decided.

romainbrenguier commented:
@owen-jones-diffblue I added some cases in the unit test, in particular some with unknown in the value set.

owen-mc-diffblue commented:
Btw failed symbols are covered by the regression test that was merged in https://github.com/diffblue/cbmc/pull/4659, I believe.

smowton commented on src/goto-symex/symex_value_set.h:
```diff
@@ -0,0 +1,74 @@
+
+/// Wrapper for value_sett which ensures we access it in a consistent way.
+/// In particular level 1 names should be used.
+class symex_value_sett : value_sett
```
Generally prefer composition to inheritence

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -222,13 +138,13 @@ renamedt<ssa_exprt, L2> goto_symex_statet::assignment(
     ssa_exprt l1_lhs(lhs);
     l1_lhs.remove_level_2();
 
-    if(run_validation_checks)
```
These should still be guarded by `if(run_validation_checks)`

tautschnig commented on src/goto-symex/renamed.cpp:
```diff
@@ -14,9 +14,6 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 bool is_l1_renamed(const exprt &expr)
 {
-  if(!is_l2_renamed(expr.type()))
-    return false;
```
The claim made in the commit message seems wrong: the type may well be renamed to L2 even when the object itself is only L1. On the contrary, it actually makes no sense for the type to be renamed to L1 ever.

romainbrenguier replied:
Ok I didn't know that the type could be L2 while the object L1. But is it true that L1 object always have L2 types? It sounds wrong to me. It seems that when an `ssa_exprt` is renamed to L1, its type is also renamed to L1.

tautschnig replied:
Renaming to L1 effectively is instantiation/determines which instance of a named object we are using. At that point the type must be complete, i.e., must be L2.

romainbrenguier replied:
It seems to be the case in symex_decl but not in symex_dereference and the other places where L1 renaming is used. Should we make `rename<L1>` apply `L2` renaming on the type to ensure this is consistent?

tautschnig commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -22,6 +22,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <langapi/language_util.h>
 #include <util/range.h>
+#include <util/format_expr.h>
+#include <util/format_type.h>
```
But then `#include <langapi/language_util.h>` better be removed as well (and hopefully also the corresponding entry in `module_dependencies.txt`).

tautschnig commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -20,10 +20,11 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/prefix.h>
 #include <util/simplify_expr.h>
 
+#include <goto-symex/renamed.h>
```
I don't think a dependency from pointer-analysis towards goto-symex is right.

tautschnig reviewed (changes requested):
This PR appears to be doing at least 3 different things (and some of them I believe are wrong, see detailed comments).

romainbrenguier commented:
@tautschnig 
> This PR appears to be doing at least 3 different things (and some of them I believe are wrong, see detailed comments).

Yes you are right, I extracted what comes after the unit tests and bugfix into a new pull request
https://github.com/diffblue/cbmc/pull/4778 so that we can merge the fix more quickly and leave time for discussion on the other part.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e26f4cd523f751cb6c3f5f6614343a28f0a8d983).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115206278

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 767d8f09ff7d519fc64c2f11b5e912050d5d3e40).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115226966
